src/OpenTelemetry/Trace/Sampler.cs

```diff
@@ -22,10 +22,19 @@ namespace OpenTelemetry.Trace
     /// </summary>
     public abstract class Sampler
     {
+        protected string description;
```

Comment:
@cijothomas curious: is it fine to put protected members before public ones?

I don't have strong opinion on such coding style thing as long as we have consistency. I was following the style here https://github.com/open-telemetry/opentelemetry-dotnet/blob/ea107d3b6982b575e40da772b75f27d5c7450ea1/src/OpenTelemetry/Trace/Samplers/ParentOrElseSampler.cs#L27.

Reply:
And should I make this a private, and expose a protected setter instead (so child classes don't see the traits)?

Reply:
I think field description should be private and the property should have a protected setter.

What if we did:
```csharp
    public abstract class Sampler
    {
        /// <summary>
         /// Gets the sampler description.
         /// </summary>
        public virtual string Description { get => this.GetType().Name; }
    }

    public sealed class ParentOrElseSampler : Sampler
    {
        private readonly Sampler delegateSampler;

        /// <summary>
        /// Initializes a new instance of the <see cref="ParentOrElseSampler"/> class.
        /// </summary>
        /// <param name="delegateSampler">The <see cref="Sampler"/> to be called to decide whether or not to sample a root trace.</param>
        public ParentOrElseSampler(Sampler delegateSampler)
        {
            this.delegateSampler = delegateSampler ?? throw new ArgumentNullException(nameof(delegateSampler));
        }

        /// <inheritdoc />
        public override string Description { get => $"ParentOrElse{{{this.delegateSampler.Description}}}"; }
    }
```

Virtual instead of abstract so if dev is fine with default they won't need to do anything?

Reply:
Will change it to protected setter.

I cannot make make `Description { get => this.GetType().Name; }` as it could introduce perf issue.
The `Description` value is quite static (unlikely to change for each call) so we want to cache it.

Reply:
Private field/protected setter works. But it doesn't have to be _literally_ that snippet. You can put a field behind it, like `public virtual string Description { get => this.descriptionCachedValue; }` to save the allocation/cycles. The virtual was the thing I was trying to highlight 😄 

I don't see Description being called anywhere? Not even sure what it is for, other than spec saying we need it.

Reply:
How about this?
```csharp
    public abstract class Sampler
    {
        protected Sampler()
        {
            this.Description = this.GetType().Name;
        }

        /// <summary>
        /// Gets or sets the sampler description.
        /// </summary>
        public string Description { get; protected set; }
    }
```
I think that will be maximum perf.

Usage:
```csharp
        public ParentOrElseSampler(Sampler delegateSampler)
        {
                this.Description = $"ParentOrElse{{{this.delegateSampler.Description}}}";
        }

        public UserSampler : Sampler
        {
                // Doesn't need to do anything.
        }
```

Reply:
> I don't see Description being called anywhere? Not even sure what it is for, other than spec saying we need it.

That's a very good question! I was asking @rajkumar-rangaraj yesterday "why would we need that, and should we do it in an idiomatic way such like ToString in C#?".

Reply:
@CodeBlanch your code seems much cleaner than mine. Stealing it now 😄